enterprise/server/tasksize/tasksize.go

```diff
@@ -64,20 +86,140 @@ func Register(env environment.Env) error {
 	return nil
 }
 
-type taskSizer struct{}
+type taskSizer struct {
+	env environment.Env
+	rdb redis.UniversalClient
+}
 
 func NewSizer(env environment.Env) (*taskSizer, error) {
-	return &taskSizer{}, nil
+	var rdb redis.UniversalClient
+	if *useMeasuredSizes {
+		rdb = env.GetRemoteExecutionRedisClient()
+		if rdb == nil {
+			return nil, status.FailedPreconditionError("missing Redis client configuration")
+		}
+	}
+	return &taskSizer{
+		env: env,
+		rdb: rdb,
+	}, nil
 }
 
 func (s *taskSizer) Estimate(ctx context.Context, task *repb.ExecutionTask) *scpb.TaskSize {
-	// TODO(bduffany): Use previous execution stats
-	return Estimate(task)
+	defaultSize := Estimate(task)
+	if !*useMeasuredSizes {
+		return defaultSize
+	}
+	recordedSize, err := s.lastRecordedSize(ctx, task)
+	if err != nil {
+		log.CtxWarningf(ctx, "Failed to read task size from Redis; falling back to default size estimate: %s", err)
+		return defaultSize
+	}
+	if recordedSize == nil {
+		return defaultSize
```

Comment:
add a //TODO that this is where we'd eventually return unsized() and let the executor run this task once to determine the size?

Reply:
Done